File: 1/2020.10.09/47.Permutations II/python/solution.py
class Solution:
    def permuteUnique(self, nums: List[int]) -> List[List[int]]:
        def backtracking(tmp):
            # terminator
            if len(tmp) == len(nums):
                res.append(tmp[:])
                return
            # process current level
            for i in range(len(nums)):
                if vis[i] or (i > 0 and nums[i] == nums[i-1] and not vis[i-1]):
                    continue
                tmp.append(nums[i])
                vis[i] = True
                # drill down
                backtracking(tmp)
                # reset states
                tmp.pop()
                vis[i] = False
        res = []
        vis = [False for _ in range(len(nums))]
        nums = sorted(nums)
        backtracking([])
        return res

File: 1/2020.10.09/47.Permutations II/java/solution.java
class Solution {
    
    List<List<Integer>> res = new ArrayList<>();
    
    public List<List<Integer>> permuteUnique(int[] nums) {
        Arrays.sort(nums);
        boolean[] vis = new boolean[nums.length];
        backtracking(nums, vis, new ArrayList<>());
        return res;
    }
    
    private void backtracking(int[] nums, boolean[] vis, List<Integer> tmp) {
        // terminator
        if (tmp.size() == nums.length) {
            res.add(new ArrayList<>(tmp));
            return;
        }
        // process current level
        for (int i = 0; i < nums.length; i++) {
            if (vis[i] || (i > 0 && nums[i] == nums[i-1] && !vis[i-1])) {
                continue;
            }
            tmp.add(nums[i]);
            vis[i] = true;
            // drill down
            backtracking(nums, vis, tmp);
            // reset states
            tmp.remove(tmp.size()-1);
            vis[i] = false;
        }
    }
}
